### PR TITLE
slipstream: init at 1.9.1

### PR DESCRIPTION
**Pull request by mibmo**
###### Description of changes

Slipstream is a mod manager for the game FTL: Faster Than Light. It runs on old java (jdk8) and is built with Maven.
Should be able to run on Darwin too (and 32-bit Linux I think) but I haven't looked into that yet.

This is my first contribution to Nix so feedback is **especially** welcome! :)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/games/slipstream/default.nix
Comment (drupol) [inline pkgs/tools/games/slipstream/default.nix:14]:
```suggestion
    hash = "sha256-F+o94Oh9qxVdfgwdmyOv+WZl1BjQuzhQWaVrAgScgIU=";
```

File: pkgs/tools/games/slipstream/default.nix
Comment (drupol) [inline pkgs/tools/games/slipstream/default.nix:6]: I think we should use the `finalAttrs` pattern here.

Since this is your first PR, let me share a couple of useful links regarding this:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/tools/games/slipstream/default.nix
Comment (mibmo) [reply pkgs/tools/games/slipstream/default.nix]: Ah, thank you! The finalAttrs pattern is a million times better! :) I've struggled with overriding the source version of other packages before so am very happy to implement this pattern - and thank you for the feedback!

**Review by drupol — CHANGES_REQUESTED** [review]
The phases are missing the hooks. You should find a lot of example in nixpkgs. Feel free to ping me for help.

**Comment by wegank** [comment]
@ofborg build slipstream slipstream.passthru.tests

File: pkgs/tools/games/slipstream/default.nix
Comment (wegank) [inline pkgs/tools/games/slipstream/default.nix:18]: `maven` should go to `nativeBuildInputs`.

File: pkgs/tools/games/slipstream/default.nix
Comment (wegank) [inline pkgs/tools/games/slipstream/default.nix:3]: Maven packages in nixpkgs are built using `maven.buildMavenPackage` (aka double invocation) these days. You might want to check [this package](https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/keycloak/scim-keycloak-user-storage-spi/default.nix) as an example.

**Comment by drupol** [comment]
I'll set this PR in `draft` until things are solved.

File: pkgs/tools/games/slipstream/default.nix
Comment (mibmo) [reply pkgs/tools/games/slipstream/default.nix]: Had to rebase from master but done :) it's not possible to do the `finalAttrs` pattern with `buildMavenPackage`, is it? only `stdenv.mkDerivation` I assume. If not, is there a better pattern than wrapping the whole derivation in `rec`?

**Comment by mibmo** [comment]
@drupol thanks for the help! :) I think I've got the hooks right now, but please do let me know if not.

File: pkgs/top-level/all-packages.nix
Comment (drupol) [inline pkgs/top-level/all-packages.nix:1786]:
```suggestion
  slipstream = callPackage ../tools/games/slipstream {
    jdk = jdk8;
  };
```

File: pkgs/tools/games/slipstream/default.nix
Comment (drupol) [inline pkgs/tools/games/slipstream/default.nix:3]:
```suggestion
```

File: pkgs/tools/games/slipstream/wrapper.sh
Comment (drupol) [inline pkgs/tools/games/slipstream/wrapper.sh:11]: Is this really necessary ?

File: pkgs/tools/games/slipstream/default.nix
Comment (drupol) [inline pkgs/tools/games/slipstream/default.nix:35]: Is this wrapper really necessary ? It looks weird.

**Review by drupol — CHANGES_REQUESTED** [review]
I made some comments.

Don't forget that your PR should only contains 2 commits in the end:

1. One commit for adding the new maintainer (you)
2. One commit for the new derivation.

File: pkgs/tools/games/slipstream/wrapper.sh
Comment (mibmo) [reply pkgs/tools/games/slipstream/wrapper.sh:11]: It's the only way I was able to get it to work; Slipstream checks if a few directories exist in the directory it's run (namely the mods directory). They can't be in the nix store since they wouldn't be writable to add mods.

I'd be 100% okay with changing the paths or doing something else; I was just trying to get it to work and judging by people debugging Slipstream online this is what I deemed best. :)

File: pkgs/top-level/all-packages.nix
Comment (mibmo) [reply pkgs/top-level/all-packages.nix:1786]: I was considering this too, but is it really necessary? I've tested Slipstream and it's both unmaintained and only works with jdk8. Or is it just a general pattern that's good to follow? I was under the impression it was mainly used for having multiple versions of a package (i.e. `something` and `something-qt`). Sorry if my questions are stupid, I hope it's okay to ask! :)

File: pkgs/tools/games/slipstream/default.nix
Comment (mibmo) [reply pkgs/tools/games/slipstream/default.nix:35]: I'm not sure of an alternative but I'm certainly open to it. I suppose it might be possible with `makeWrapper`?

File: pkgs/tools/games/slipstream/wrapper.sh
Comment (gaelreyrol) [inline pkgs/tools/games/slipstream/wrapper.sh:3]: I would rather use `$XDG_DATA_HOME` instead of `$HOME` to not override user's configuration as it could be elsewhere.
```suggestion
_base="$XDG_DATA_HOME/slipstream"
```

File: pkgs/tools/games/slipstream/wrapper.sh
Comment (gaelreyrol) [inline pkgs/tools/games/slipstream/wrapper.sh:5]:
```suggestion
mkdir -p "$_base/{mods,backup}"
```

File: pkgs/tools/games/slipstream/wrapper.sh
Comment (gaelreyrol) [reply pkgs/tools/games/slipstream/wrapper.sh:11]: It seems Slipstream is indeed working like this. Maybe you could add a comment to explain / point a line in the source code? Looks like it is around here: https://github.com/Vhati/Slipstream-Mod-Manager/blob/master/src/main/java/net/vhati/modmanager/cli/SlipstreamCLI.java#L105

File: pkgs/tools/games/slipstream/wrapper.sh
Comment (gaelreyrol) [inline pkgs/tools/games/slipstream/wrapper.sh:1]: I believe you should not specify a shebang. @drupol wdyt?

File: pkgs/tools/games/slipstream/default.nix
Comment (gaelreyrol) [reply pkgs/tools/games/slipstream/default.nix:35]: I do not feel comfortable with this as well, I would find it proper if embedded in the derivation:
```nix
{
  let
    wrapper = ./wrapper.sh
  in mkDerivation {
    installPhase = ''
      makeWrapper ${wrapper}
    '';
  }
}
```

File: pkgs/tools/games/slipstream/wrapper.sh
Comment (gaelreyrol) [reply pkgs/tools/games/slipstream/wrapper.sh:1]: Or use this `#! @shell@` and provide replace it with `substitute --subst-var-by shell stdenv.shell`

Comment (drupol) [reply pkgs/tools/games/slipstream/wrapper.sh:1]: I don't know. @pennae suggested to use `makeShellWrapper`.

Comment (gaelreyrol) [reply pkgs/tools/games/slipstream/wrapper.sh:1]: makeWrapper is almost the same as `makeShellWrapper`.

Comment (drupol) [reply pkgs/tools/games/slipstream/wrapper.sh:1]: I'm totally useless here, I don't know what's the best option to be honest :disappointed:

Comment (gaelreyrol) [reply pkgs/tools/games/slipstream/wrapper.sh:1]: > I'm totally useless here, I don't know what's the best option to be honest disappointed

No you aren't :hugs:. I feel like `wrapper.sh` is not needed and as you suggested it can be done with multiple `--run` options.

File: pkgs/tools/games/slipstream/default.nix
Comment (gaelreyrol) [inline pkgs/tools/games/slipstream/default.nix:30]: This would remove the need for a `wrapper.sh` file:
```suggestion
    makeWrapper ${wrapper} $out/bin/${pname} \
      --suffix PATH : ${lib.makeBinPath [ jdk ]} \
      --run 'cd $XDG_DATA_HOME/slipstream'
      --run 'mkdir -p $_base/{mods,backup}'
      --run 'java -jar $out/share/java/modman.jar'
```

File: pkgs/tools/games/slipstream/wrapper.sh
Comment (mibmo) [reply pkgs/tools/games/slipstream/wrapper.sh:3]: Good point :) is it reliable enough to depend on XDG_DATA_HOME or should I perhaps write some fallback logic?

Comment (gaelreyrol) [reply pkgs/tools/games/slipstream/wrapper.sh:3]: I think it is reliable enough because it requires a desktop environment if I understand correctly, maybe it could be asserted but I don't know how right now ;)

File: pkgs/tools/games/slipstream/wrapper.sh
Comment (mibmo) [reply pkgs/tools/games/slipstream/wrapper.sh:1]: Since we're removing the wrapper anyway I'll resolve this, but thanks for discussing it :)

**Comment by mibmo** [comment]
Thanks everyone for the feedback. :)
I've implemented most anything, but I do have one point of confusion; since we're now using `buildMavenPackage` to build the derivation, we can't use the `finalAttrs` pattern (as with `stdenv.mkDerivaiton`). I've reverted to using `rec` but I'm wondering if perhaps there's a better way?

> I made some comments.
> 
> Don't forget that your PR should only contains 2 commits in the end:
>     1. One commit for adding the new maintainer (you)
>     2. One commit for the new derivation.
Sgtm. I figure I'll just squash them when I get the go-ahead :)

File: pkgs/top-level/all-packages.nix
Comment (mibmo) [reply pkgs/top-level/all-packages.nix:1786]: I've decided to do it anyhow - I'm guessing it's because it makes it easier to override? (read: possible ...). Not sure how I missed that but thank you anyway :)

File: pkgs/tools/games/slipstream/default.nix
Comment (mibmo) [reply pkgs/tools/games/slipstream/default.nix]: Implemented this & fallback logic in case `XDG_DATA_DIR` is undefined.

**Comment by drupol** [comment]
Hi,

Thanks for implementing the changes.
However, this shouldn't have been reverted to using `rec`.

What was the issue that makes you revert it ?

**Comment by wegank** [comment]
@drupol Custom builders generally don't support function-based attribute recursion. For example, there isn't a single `buildRustPackage (finalAttrs:` line in nixpkgs yet.

**Comment by drupol** [comment]
> @drupol Custom builders generally don't support function-based attribute recursion. For example, there isn't a single `buildRustPackage (finalAttrs:` line in nixpkgs yet.

Oops, my bad. Please ignore my previous comment then.

File: pkgs/tools/games/slipstream/default.nix
Comment (gaelreyrol) [inline pkgs/tools/games/slipstream/default.nix:34]: Did you forget to remove this?

File: pkgs/tools/games/slipstream/default.nix
Comment (gaelreyrol) [inline pkgs/tools/games/slipstream/default.nix:43]: Can you specify at least one platform?

File: pkgs/tools/games/slipstream/default.nix
Comment (mibmo) [reply pkgs/tools/games/slipstream/default.nix:34]: ? Nope. Sorry, I'm a bit confused, but how would it be possible to run this without `makeWrapper`? I thought this was in the same vein as what you suggested :)

Comment (gaelreyrol) [reply pkgs/tools/games/slipstream/default.nix:34]: I am only speaking about this line: `--run 'echo $XDG_DATA_HOME > $HOME/debug' \`

File: pkgs/tools/games/slipstream/default.nix
Comment (mibmo) [reply pkgs/tools/games/slipstream/default.nix:43]: Sure! :) I'm adding `platforms.all` since it's Java, but feel free to correct if that's wrong.

File: pkgs/tools/games/slipstream/default.nix
Comment (mibmo) [reply pkgs/tools/games/slipstream/default.nix:34]: ohhh, my mistake. Yep, totally :facepalm: thanks for catching that!

**Comment by drupol** [comment]
What's the status of this PR ?

**Comment by mibmo** [comment]
> What's the status of this PR ?

Waiting on reviews & TODO is also to squash the commits :)

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
@GrahamcOfBorg build slipstream

**Pull request merged**
